### PR TITLE
[cxx-interop] Don't expose async or @_alwaysEmitIntoClient functions

### PR DESCRIPTION
**Pull request by WANGJIEKE**
Ensure that top-level `async` functions are not exposed to C/C++ because there's currently no support for Swift concurrency being imported into C/C++.

Also, don't expose transparent functions (marked with `@_transparent` attribute) since it's not supported on the clang side.

### CODE REVIEW

**Comment by WANGJIEKE** [comment]
@hyp Could you please review this? Thanks. 😀

**Comment by slavapestov** [comment]
@_transparent functions still have an ABI entry point though, so it should be fine to call them like an ordinary function. @_alwaysEmitIntoClient is the one that's not ABI.

**Comment by slavapestov** [comment]
Also ConstructorDecl can also be async, too.

**Comment by hyp** [comment]
@WANGJIEKE Based on Slava's feedback we should allow `@_transparent` functions to be called. Let's have a test where they can be called from C++. However, we should not allow the calls to `@_alwaysEmitIntoClient` functions, so let's prohibit those.

**Comment by hyp** [comment]
> Also ConstructorDecl can also be async, too.

We don't emit constructors in C++ yet. We will have to prohibit async constructs though when we add support for them, thanks for bringing it up.

**Comment by WANGJIEKE** [comment]
Updated. I will also rename the pull request to reflect this change.

**Comment by hyp** [comment]
@swift-ci please test

**Comment by WANGJIEKE** [comment]
Looks like the test failed because it's running on a version of macOS that doesn't support concurrency. I'm thinking separating this test into two so that we can still test `@_alwaysEmitIntoClient` with as many versions of macOS as possible.

For the `async` test, after examining through lit documentation and config files, my guess is we can use `REQUIRES: concurrency` to make sure it only runs in supported platforms.

Is this a viable solution to the test failure?

**Comment by hyp** [comment]
@swift-ci please test

**Comment by WANGJIEKE** [comment]
Looks like only adding `REQUIRES: concurrency` didn't work 😐
I'm thinking of adding `REQUIRES: concurrency_runtime` as well.

@hyp What do you think?

**Comment by WANGJIEKE** [comment]
Updated :)

**Comment by hyp** [comment]
@swift-ci please test

File: lib/PrintAsClang/DeclAndTypePrinter.cpp [hunk 1]
```diff
@@ -1110,6 +1110,12 @@ class DeclAndTypePrinter::Implementation
 
   void visitFuncDecl(FuncDecl *FD) {
     if (outputLang == OutputLanguageMode::Cxx) {
+      // Don't expose async functions or @_alwaysEmitIntoClient functions
+      // because they're currently unsupported
+      if (FD->hasAsync() || FD->getAttrs().hasAttribute<AlwaysEmitIntoClientAttr>()) {
```
Comment (hyp) [inline lib/PrintAsClang/DeclAndTypePrinter.cpp:1115]: One final NIT: Please format this line to follow the 80 columns style limit. For example you can run something like this:

```
git diff -U0 --no-prefix HEAD~1 HEAD -- lib/ include/ | ~/llvm-project/clang/tools/clang-format/clang-format-diff.py -style LLVM -i -binary $(xcrun --find clang-format)
```

**Review by hyp — APPROVED** [review]
LGTM, with one final request

File: lib/PrintAsClang/DeclAndTypePrinter.cpp (hunk 1 above)
Comment (WANGJIEKE) [reply lib/PrintAsClang/DeclAndTypePrinter.cpp:1115]: Somehow I didn't get any output from this command, and `git status` shows there's no change.

Here's my version of this command:

```bash
git diff -U0 --no-prefix HEAD~1 HEAD -- lib/ include/ | ../llvm-project/clang/tools/clang-format/clang-format-diff.py -style LLVM -i -binary $(xcrun --find clang-format)
```

And it's run inside `swift-project/swift`.

Comment (WANGJIEKE) [reply lib/PrintAsClang/DeclAndTypePrinter.cpp:1115]: I managed to manually invoke clang-format and I'm going to push the commit.

Comment (WANGJIEKE) [reply lib/PrintAsClang/DeclAndTypePrinter.cpp]: I think I know why it doesn't work. `HEAD~1` didn't change DeclAndTypePrinter.cpp so there's no output from it.

**Comment by WANGJIEKE** [comment]
I think it's time to add something like "format on save" to my editor 😀

**Comment by hyp** [comment]
@swift-ci please test

**Comment by hyp** [comment]
@swift-ci please test source compatibility

**Comment by hyp** [comment]
`FAIL: async-http-client, 5.0, a72c5a, Swift Package` is unrelated, going to merge.

**Pull request merged**